scipy/signal/_peak_finding.py

```diff
@@ -337,77 +346,38 @@ def peak_prominences(x, peaks, wlen=None):
     >>> plt.vlines(x=peaks, ymin=contour_heights, ymax=x[peaks])
     >>> plt.show()
     """
-    x = np.asarray(x)
-    peaks = np.asarray(peaks)
-    if peaks.size == 0:
-        # Handle empty peaks
-        return np.array([]), np.array([], dtype=int), np.array([], dtype=int)
-
+    # Inner function expects `x` to be contiguous
+    x = np.ascontiguousarray(x, dtype=np.float64)
     if x.ndim != 1:
         raise ValueError('`x` must have exactly one dimension')
+
+    peaks = np.asarray(peaks)
     if peaks.ndim != 1:
         raise ValueError('`peaks` must have exactly one dimension')
-    if x.size <= peaks.max():
-        raise ValueError('an index in `peaks` exceeds the size of `x`')
-    if not np.issubdtype(peaks.dtype, np.integer):
-        raise ValueError('`peaks` must be an array of integers')
-    if wlen is not None and wlen < 3:
-        raise ValueError('`wlen` must be at least 3')
-
-    # Prepare return arguments
-    prominences = np.zeros(peaks.size)
-    left_bases = np.zeros(peaks.size, dtype=int)
-    right_bases = np.zeros(peaks.size, dtype=int)
-
-    for i, peak in enumerate(peaks):
-        # If wlen is twice the size of x the symmetric window always covers
-        # the full x
-        if wlen is not None and wlen < x.size * 2:
-            wlen = int(wlen)
-            # Calculate window borders around the evaluated peak
-            wleft = peak - wlen // 2
-            wright = peak + wlen // 2
-            # Handle border cases
-            wleft = 0 if wleft < 0 else wleft
-            wright = x.size if x.size < wright else wright
-            # Use slice for prominence calculation
-            window = x[wleft:wright]
-            # Correct peak position in x
-            peak -= wleft
-        else:
-            # Use full x for prominence calculation
-            window = x
-            wleft = 0
-
-        # Positions where window is larger than current peak height
-        greater_peak = np.where(window > window[peak])[0]
-
-        try:
-            # Nearest position to the left of peak with
-            # window[left] > window[peak]
-            left = greater_peak[greater_peak < peak].max()
-        except ValueError:
-            left = 0
+    if peaks.size == 0:
+        # The astype statement below would raise a TypeError for empty arrays
+        # which default to dtype np.float64 thus handle this case separately
+        peaks = np.array([], dtype=np.intp)
```

Comment:
I'm not to happy with the verbose input validation this function has right now. E.g. the handling of `peaks` above. Do you guys see a way to simplify this? Is the complexity in this case worth it?

My thoughts where that in Python there is always the option to be very liberal with the input arguments. However I think user-facing code should fail early and with descriptive error messages. Basically fail for everything instead of converting implicitly.

Reply:
Question: is there a better way better way to check if `peaks` can be safely converted to `np.intp`? That would make it possible to simply use `np.asarray(peaks, order='C', dtype=np.intp)` if I new that type conversion would be safe.

Reply:
I think this is an okay way to do it. I don't immediately see a better one.